minio/minio
# azure gateway: return MD5Sum as ETag for S3 API compatibility

Fixes #6872.



ETag generated in PutObjectPart is a UUID string rather than the MD5Sum of data being uploaded for Azure gateway. This poses a problem when edge caching mode is on, because part ETag validation is done at the time of CompleteMultipartUpload and it is expected that ETag be a hex encoded string.
## Description

Since the Azure BlockBlob api does not allow setting metadata, save a temporary part metadata file with the association between the uuid based ETag (used as part of the block id) and the md5sum which is known after data is written to backend. This temporary part metadata is cleaned up on completion of upload or aborting of upload. This allows PutObjectPart to be [S3 compatible](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTCommonResponseHeaders.html) w.r.t etag
## Motivation and Context


allows objects uploaded to Azure via multipart API to be saved in edge cache when edge cache is enabled.
## Regression
 No


## How Has This Been Tested?



1. enable edge cache ```export MINIO_CACHE_DRIVES=/tmp/testcache```
2. start azure gateway ./minio gateway azure
3. mc cp /tmp/large myazure/testbucket/multipart
4.  You should see testbucket/multipart cached in /tmp/testcache folder
## Types of changes


## Checklist:
